text/0000-futures.md

```diff
@@ -0,0 +1,520 @@
+impl LocalWaker {
+    /// Wake up the task associated with this `LocalWaker`.
+    pub fn wake(&self);
+}
+
+/// You can upgrade to a sendable `Waker` at zero cost, but waking through a `Waker` is more expensive
+/// due to synchronization.
+impl From<LocalWaker> for Waker  { .. }
+```
+
+Task execution always happens in the context of a `LocalWaker` that can be used to
+wake the task up locally, or converted into a `Waker` that can be sent to other threads.
+
+It's possible to construct a `Waker` using `From<Arc<dyn Wake>>`.
+
+### `UnsafeWake` and `no_std` compatibility
+
+The [`UnsafeWake` trait](https://doc.rust-lang.org/nightly/std/task/trait.UnsafeWake.html)
+in `core::task` is designed to support task wakeup in  `no_std` environments, where
+we cannot use `Arc`. It is *not* proposed for stabilization at this time, because
+its APIs are awaiting revision based on object safety for `*mut self` methods.
+
+## `core::future` module
+
+With all of the above task infrastructure in place, defining `Future` is
+straightforward:
+
+```rust
+pub trait Future {
+    /// The type of value produced on completion.
+    type Output;
+
+    /// Attempt to resolve the future to a final value, registering
+    /// the current task for wakeup if the value is not yet available.
+    ///
+    /// # Return value
+    ///
+    /// This function returns:
+    ///
+    /// - [`Poll::Pending`] if the future is not ready yet
+    /// - [`Poll::Ready(val)`] with the result `val` of this future if it
+    ///   finished successfully.
+    ///
+    /// Once a future has finished, clients should not `poll` it again.
+    ///
+    /// When a future is not ready yet, `poll` returns `Poll::Pending` and
+    /// stores a clone of the [`LocalWaker`] to be woken once the future can
+    /// make progress. For example, a future waiting for a socket to become
+    /// readable would call `.clone()` on the [`LocalWaker`] and store it.
+    /// When a signal arrives elsewhere indicating that the socket is readable,
+    /// `[LocalWaker::wake]` is called and the socket future's task is awoken.
+    /// Once a task has been woken up, it should attempt to `poll` the future
+    /// again, which may or may not produce a final value.
+    ///
+    /// Note that on multiple calls to `poll`, only the most recent
+    /// [`LocalWaker`] passed to `poll` should be scheduled to receive a
+    /// wakeup.
+    ///
+    /// # Runtime characteristics
+    ///
+    /// Futures alone are *inert*; they must be *actively* `poll`ed to make
+    /// progress, meaning that each time the current task is woken up, it should
+    /// actively re-`poll` pending futures that it still has an interest in.
+    ///
+    /// The `poll` function is not called repeatedly in a tight loop-- instead,
+    /// it should only be called when the future indicates that it is ready to
+    /// make progress (by calling `wake()`). If you're familiar with the
```

Comment:
This seems to contradict the previous paragraph: Futures are inert, but they have to call `wake()` to make progress? How are they supposed to do that if they are inert?

Reply:
The future itself does not typically call `wake()` itself. Instead, it stores a `Local/Waker` somewhere for someone else to call. For example, in a channel, the receiving stream will look to get a next element. If it doesn't see one available, it'll store a `Waker` so that the next writer that comes along and adds an element to the channel can call `wake()`, causing the listener to be woken up and see the new element.

Reply:
I had a similar confusion. Trying to unpick why: "the future indicates that it is ready to make progress (by calling `wake()`)" seems imply that the future actively indicates it is ready to make progress by calling wake. When in fact it's more general, e.g. "something has called `wake()` in order to indicate the future is ready to make progress"